src/Backend/Haskell.idr

```diff
@@ -78,55 +83,56 @@ renderDef (ADT     decl cases) = text "data" |++| renderDecl decl
-makeType e td@(TName name _) = HsParam name . map HsVar $ getFreeVars (getUsedVars e td)
-
-||| Generate Haskell type definitions from a `TDef`, including all of its dependencies.
-makeDefs : Env n -> TDef n -> State (List Name) (List Haskell)
-makeDefs _ T0            = pure []
-makeDefs _ T1            = pure []
-makeDefs e (TProd xs)    = map concat $ traverse (assert_total $ makeDefs e) xs
-makeDefs e (TSum xs)     = map concat $ traverse (assert_total $ makeDefs e) xs
-makeDefs _ (TVar v)      = pure []
-makeDefs e td@(TMu name cs) =
-   do st <- get
-      if List.elem name st then pure []
-       else let
-          decl = MkDecl name (getFreeVars (getUsedVars e td))
-          newEnv = Right decl :: e
-          args = map (map (makeType newEnv)) cs
-         in
-        do res <- map concat $ traverse {b=List Haskell} (\(_, bdy) => assert_total $ makeDefs newEnv bdy) (toList cs)
-           put (name :: st)
-           pure $ ADT decl args :: res
-makeDefs e td@(TName name body) =
-  do st <- get
-     if List.elem name st then pure []
-       else
-        do res <- assert_total $ makeDefs e body
-           put (name :: st)
-           pure $ Synonym (MkDecl name $ getFreeVars (getUsedVars e td)) (makeType e body) :: res
-
-Backend Haskell where
-  generateTyDefs e td = reverse $ evalState (makeDefs e td) []
-  generateCode        = renderDef
-  freshEnv            = freshEnvLC
-
-NewBackend Haskell HsType where
-  msgType          = makeType (freshEnv {lang=Haskell} 0)
-  typedefs td      = reverse $ evalState (makeDefs (freshEnv {lang=Haskell} 0) td) []
-  source type defs = vsep2 $ map renderDef $ Synonym (MkDecl "TypedefSchema" []) type :: defs 
-
-||| Generate type body, only useful for anonymous tdefs (i.e. without wrapping Mu/Name)
-generateType : TDef n -> Doc
-generateType {n} = renderType . makeType (freshEnv {lang=Haskell} n)
+makeType _    T0          = HsVoid
+makeType _    T1          = HsUnit
+makeType e    (TSum xs)   = foldr1' (\t1,t2 => HsParam "Either" [t1, t2]) (map (assert_total $ makeType e) xs)
+makeType e    (TProd xs)  = HsTuple $ map (assert_total $ makeType e) xs
+makeType e    (TVar v)    = either HsVar hsParam $ Vect.index v e
+makeType e td@(TMu cases) = HsParam (nameMu cases) . map (either HsVar hsParam) $ getUsedVars e td
+makeType e    (TApp f xs) = HsParam (name f) (map (assert_total $ makeType e) xs)
+
+||| Generate a Haskell type from a `TNamed`.
+makeType' : Env n -> TNamed n -> HsType
+makeType' e (TName name body) = HsParam name . map (either HsVar hsParam) $ getUsedVars e body
+
+mutual
+  ||| Generate all the Haskell type definitions that a `TDef` depends on.
+  makeDefs : TDef n -> State (List Name) (List Haskell)
+  makeDefs    T0          = pure []
+  makeDefs    T1          = pure []
+  makeDefs    (TProd xs)  = map concat $ traverse (assert_total makeDefs) xs
+  makeDefs    (TSum xs)   = map concat $ traverse (assert_total makeDefs) xs
```

Comment:
`concat <$> traverse (assert_total makeDefs) xs` to make it consistent with `TApp` case?